docs: updates

Berkmann18 opened the pull request:
- Level 1-3 diagrams and some notes.
- GitHub Action and Netlify setup.

Berkmann18 committed:
docs: added the core information

Berkmann18 commented:
@gr2m What do you think about this?

Berkmann18 committed:
docs: clean-up and added a new set of diagrams

2 architectural ideas (a lot of progress left)

gr2m commented on .czrc:
```diff
@@ -0,0 +1 @@
+{ "path": "cz-conventional-changelog" }
```
do we need this?

Berkmann18 replied:
Not particularly, it could go in the `.gitignore` file, I just find it easier as I use `commitizen` which is set to ESLint (except for individual projects where the Angular commit convention is used).

gr2m replied:
There is a price to add this dotfile. I don't think we need to enforce commit conventions for a docs-only repository. Can you move the file further up your directory tree?

Berkmann18 replied:
That's a good idea, I usually add the CZ config (although only the relevant `devDep`) as I think it's a valuable tool to use and would prevent others from having to set up the repo-specific config and a good way to prevent people from not respecting the commit convention but you're right.

> Can you move the file further up your directory tree?

If I restructure my list of projects, that may work.

gr2m replied:
> a good way to prevent people from not respecting the commit convention

Since GitHub introduced the "squash & merge" button, I no longer enforce the commit messages, I just make sure that the maintainers who review & merge pull requests are aware of them. It's one thing less we need to bother new contributors with

gr2m commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+.idea
+.vscode
```
do we need this?

Berkmann18 replied:
I would be keen on not ignoring the VSCode folder but I don't want to impose the extra files on others who use other editors/IDE.

gr2m replied:
maybe add that line to a global gitignore file? That's what I did. I don't mind adding the file, I just think there is value in having less dotfiles if we can help it, and there are no downsides to new contributors.

Berkmann18 replied:
That's a good idea. It's usually part of the language-specific ignore list hence why I added that.

gr2m commented:
I'd say the website does not use the library, it will be using the specification directly.

Could you maybe add a `CONTRIBUTING.md` file with an explanation for the `.puml` files, in case someone would like to contribute to them in future?

Berkmann18 commented:
> Could you maybe add a CONTRIBUTING.md file with an explanation for the .puml files, in case someone would like to contribute to them in future?

Sure, I'll add it once I've got more lower-level diagram which I think are important (especially since the CLI itself would significantly change with the introduction of the library and co).

Berkmann18 committed:
docs: wip

Berkmann18 committed:
chore: dotfile clean-up

Berkmann18 committed:
docs: updated diagrams and added more for idea 1

Berkmann18 committed:
docs(readme): revamped the READMEs

By applying some UX principles to improve the readability and structure of the readmes

Berkmann18 committed:
docs(contributing): added necessary info

Berkmann18 commented:
I've added a contribution guide and made some changes to various diagrams, I've focused on the `idea1` ones as I thought it would be more suitable.

Berkmann18 commented on README.md:
```diff
@@ -1,2 +1,67 @@
+_**Notes**:_
+-  The diagrams use the [C4 model](https://c4model.com/#CoreDiagrams).
+-  Each containers (in the [C4 sense](https://c4model.com/#faq)) have their own repo (i.e. `1 container = 1 repo`).
+- You need to have [PlantUML](https://plantuml.com/starting) installed to be able to generate the images and preview the diagrams (this can be done via IDE/editor extension too).
```
Note to self: move this to `CONTRIBUTING.md`

Berkmann18 commented on README.md:
```diff
@@ -1,2 +1,67 @@
+This is a (potentially temporary) place where architectural diagrams of all the AC projects can be found.
+
+It's currently focused on the _new world_ (revamped) projects so it may not accurately illustrate the current systems.
+
```
```suggestion

```

Berkmann18 commented on README.md:
```diff
@@ -1,2 +1,67 @@
+      View component diagram
+    </summary>
+    <img src="" alt="Website_Component_Diagram">
+  </details>
```
I'm wondering if moving this into a `idea0` folder and linking both ideas here would be better until we've decided on the architecture at least to the container level.
Edit: I'm considering trying https://github.com/adrianvlupu/C4-Builder to have a potentially easier time managing and maintaining the docs and diagrams (which would possibly be more accessible than opening the SVGs or having HTML+SVG instead of Markdown with images or using Structurizr).

Berkmann18 commented on idea1/website.puml:
```diff
@@ -0,0 +1,30 @@
+Container_Boundary(website, "Website") {
+  Component(site, "Website", "Docusaurus", "Documentation", "html5")
+  Component(docs, "Docs", "Markdown", "Specification/CLI/Bot documentation", "")
+
```
Note to self: Add the missing relationships (when I come back to that) such as Crowdin, the docs/site link and the actual relationships between the components and the containers.

Berkmann18 committed:
fix: typo fix, re-organisation and hyperlinks

Berkmann18 committed:
docs: slight improvement

Berkmann18 committed:
feat(idea1): improved the website diagram

Berkmann18 committed:
feat(idea1): added the app diagram

Berkmann18 committed:
docs: updated both readmes

One due to a forgotten typo fix

Berkmann18 commented:
@gr2m Which of the ideas makes more sense to you so far?

gr2m commented:
I like the separation of the "core" module and the specification. In my understanding, the "core" module would have the logic to read & update the `.all-contributorsrc` file. The specification would be just data. The Specification will be useful without the core module, e.g. for the website

I like this All Contributors diagram https://github.com/all-contributors/architecture/blob/78be65cac62a2325f904a703e37d4cf888a42fe5/out/ac_container/AC_Container_Diagram.png

I'm a bit worried that if we go too much into detail it will get hard to maintain and get out of sync with the actual system

Berkmann18 commented:
Well, I have no plans in documenting all the way to C4 (Class diagram), the component diagrams (C3) would be as far as I go.
So I'll park idea 1 and simplify the documentation hopefully this weekend if work doesn't get overreaching.

Berkmann18 committed:
feat(idea1): removed the idea 1

Berkmann18 commented:
Quick update: I've been looking into using Structurizr instead which promise to make the creation and generation of all those diagrams easier to make but I've been experiencing issues (getting weird/mish-mashed diagrams and not having the C4-PlantUML export option, cf. https://github.com/structurizr/cli/issues/29) and I'm hopping to get at least a set of clearer diagrams (which could then be manually improved and maintained) but that has been trickier than expected (especially with external situations that has taken over my time, like finding a place, regular medical checks, ...).
The Structurizr stuff are seating in the `stzr-dsl` branch.

gr2m commented:
another alternative is https://excalidraw.com/. It does not allow to generate charts, but it's a great editing experience and the outcome looks great. I'm working on a GitHub integration with them right now

Berkmann18 commented:
That's true, there's also ArchiMate, C4-PlantUML is fine but when it comes to having several levels, not repeating stuff is hard.

Berkmann18 committed:
feat(diagrams): global structurizr-driven update + clean-up

Updated all the C4 diagrams based on improved versions of the C4-PlantUML exports from Structurizr
(cf. `stzr-dsl` branch). And I removed the idea1 stuff

Berkmann18 committed:
docs: update

Berkmann18 commented:
I've finally gotten around to updating and adding everything (**except** for the components in the Library and Specification containers where I'm not sure).
Let me know what you think and what you see going in the library and specification containers.

Also, do you think it would be a good idea to make this repo public or do you feel that the diagrams and docs should be moved somewhere else (e.g. the website repo) or only after the re-architecting is done?

~Another thing I'm hoping to do (although that may change if this gets integrated into https://github.com/all-contributors/all-contributors even if it's just the SVGs which are more linked to each other) is to setup a GH Action that would automatically (re-)generate the PNG/SVG versions of the diagrams on `push`.~ (I've opted for using `c4builder` with a GitHub Action built for it that generates a static site with the docs and diagrams).

Berkmann18 committed:
feat: c4built docs

Berkmann18 committed:
feat(docs): local image generation + deployment diagram emptied

Berkmann18 committed:
feat: moved to a c4builder structure

Instead of having the folder structure I went with, I'm going for a `c4builder`-led structure for
better clarity, maintainability and ease

Berkmann18 committed:
build(action): added a GitHub Action

Berkmann18 committed:
docs: updated README and CONTRIBUTING

Berkmann18 committed:
feat: clean-up and diagram improvements

Berkmann18 committed:
docs: slightly improved the docs

gr2m commented:
> Also, do you think it would be a good idea to make this repo public or do you feel that the diagrams and docs should be moved somewhere else (e.g. the website repo) or only after the re-architecting is done?

Definitely public! It's great for onboarding new contributors and maintainers

>  Let me know what you think and what you see going in the library and specification containers.

I'm not familiar with the C4 Model, probably most people won't. 

I wonder if you could rephrase [the new README](https://github.com/all-contributors/architecture/tree/dev) to be less about the abstractions you used, and more about all contributors? I don't think it should explain the abstraction as the main section of the README. I think you should show the architecture for all contributors right away, and then maybe explain abstractions you used as necessary, using aspects All Contributors as concrete example, instead of mentioning technology that we don't use, such as Apache Tomcat.

I feel like you the current README.md and HOME.md explain the Architecture Design Abstraction you are using, instead of explaining All Contributors? It's great to explain it as many people won't be familiar with it, but I think the architecture diagrams should be as self-explanatory as possible and not require understanding of the model you used.

I feel like we should put your main diagram from https://github.com/all-contributors/architecture/blob/dev/docs/1%20All%20Contributors%20System/HOME.md in the repositories README, explain the main components within the All Contributors System, and then link to resources with more information about the Architecture Design Model used

What do you think?

Berkmann18 commented:
> I wonder if you could rephrase the new README to be less about the abstractions you used, and more about all contributors? I don't think it should explain the abstraction as the main section of the README.

Good point, that was from the automatically generated README and I wasn't too fussed about removing it but I agree and I removed that section (since people can get the same info from c4model.com anyway).
Everything in `./docs` is automatically generated based on what is in `src` (I suspect it won't be even needed in the repo itself once the Netlify deployment actually works), I'm not sure if `CONTRIBUTING.md` makes that clear enough (or perhaps the folder names should be rethought).

As for the text under the `Level ...` headings, that was automatically added and I feel like keeping them would be good for those who are new to the site and the C4 model but useless for those who already know that and where a description about the underlying system/container (e.g. what the app is for, what it can do) would be preferable.
And as you rightly said, most people reading those docs probably aren't familiar with the C4 model so I'm compelled to leave those paragraphs as they are and try to get the diagrams to speak for themselves while being as clean and clear as possible.

Berkmann18 committed:
docs(readme): improved the readme

Berkmann18 committed:
chore(all-contributors): added All Contributors table

Berkmann18 committed:
build(hosting): added the Netlify config

Berkmann18 committed:
build(hosting): setup Netlify

Berkmann18 committed:
erge branch 'dev' of github.com:all-contributors/architecture into dev

Berkmann18 committed:
feat(diagrams): lib/spec diagrams [wip]

Added the components to both (previously empty) diagrams. I will add the connections later

Berkmann18 committed:
feat(diagrams): improved the presentation of some diagrams

gr2m reviewed (approved):
Alright, if you are ready from your side, I'd say let's merge it in, it's a great foundation to iterate on. Great work Maximilian!
